console/v1/types_console_sample.go

```diff
@@ -0,0 +1,183 @@
+	//
+	// It is required and must be no more than 50 characters in length.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=50
+	Title string `json:"title"`
+
+	// abstract is a short teaser the sample.
+	//
+	// It is required and must be no more than 100 characters in length.
+	//
+	// Currently, the abstract is shown on the sample card tile below the title and provider
+	// and is limited to three lines of content.
+	// +required
+	// +kubebuilder:validation:MaxLength=100
+	Abstract string `json:"abstract"`
+
+	// description of the sample.
+	//
+	// It is required and can have a maximum length of **4096** characters.
+	//
+	// It is a README.md-like content for additional information, links, pre-conditions, and other instructions.
+	// It will be rendered as Markdown so that it can contain line breaks, links, and other simple formatting.
+	// +required
+	// +kubebuilder:validation:MaxLength=4096
+	Description string `json:"description"`
+
+	// icon is an optional base64 encoded image and shown beside the sample title.
+	//
+	// The format must follow the [data: URL format].
+	//
+	//   data:[<mediatype>][;base64],<base64 encoded image>
+	//
+	// For example:
+	//
+	//   data:image;base64,             plus the base64 encoded image.
+	//
+	// Vector images can also be used. SVG icons must start with:
+	//
+	//   data:image/svg+xml;base64,     plus the base64 encoded SVG image.
+	//
+	// All sample catalog icons will be shown on a white background (also when the dark theme is used).
+	// The web console ensures that different aspect radios work correctly.
+	// Currently, the surface of the icon is at most 40x100px.
+	//
+	// [data: URL format]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs
+	// +optional
+	// +kubebuilder:validation:Pattern=`^data:`
+	Icon string `json:"icon"`
+
+	// type is an optional label to group multiple samples.
+	//
+	// It is optional and must be no more than 20 characters in length.
+	//
+	// Recommendation is a singular term like "Builder Image", "Devfile" or "Serverless Function".
+	//
+	// Currently, the type is shown a badge on the sample card tile in the top right corner.
+	// +optional
+	// +kubebuilder:validation:MaxLength=20
+	Type string `json:"type"`
+
+	// provider is an optional label to honor who provides the sample.
+	//
+	// It is optional and must be no more than 20 characters in length.
+	//
+	// A provider can be a company like "Red Hat" or an organization like "CNCF" or "Knative".
+	//
+	// Currently, the provider is only shown on the sample card tile below the title with the prefix "Provided by "
+	// +optional
+	// +kubebuilder:validation:MaxLength=20
+	Provider string `json:"provider"`
+
+	// tags like "Java", "Quarkus", etc.
+	// Tags are optional and can be used by the user to find samples in the samples catalog.
+	// They will be also displayed on the samples details page.
+	// +optional
+	Tags []string `json:"tags"`
```

Comment:
Both make sense, I will limit it to 10 tags.

Reply:
@JoelSpeed the CI job fails when testing the CRD with `// +kubebuilder:validation:UniqueItems:=true`:

```
  [38;5;9m[FAILED] Unexpected error:
      <*fmt.wrapError | 0xc0019d1800>: {
          msg: "unable to create CRD instances: unable to create CRD \"consolesamples.console.openshift.io\": CustomResourceDefinition.apiextensions.k8s.io \"consolesamples.console.openshift.io\" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[tags].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic",
          err: <*fmt.wrapError | 0xc0019d17e0>{
              msg: "unable to create CRD \"consolesamples.console.openshift.io\": CustomResourceDefinition.apiextensions.k8s.io \"consolesamples.console.openshift.io\" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[tags].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic",
              err: <*errors.StatusError | 0xc00125b860>{
                  ErrStatus: {
                      TypeMeta: {Kind: "", APIVersion: ""},
                      ListMeta: {
                          SelfLink: "",
                          ResourceVersion: "",
                          Continue: "",
                          RemainingItemCount: nil,
                      },
                      Status: "Failure",
                      Message: "CustomResourceDefinition.apiextensions.k8s.io \"consolesamples.console.openshift.io\" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[tags].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic",
                      Reason: "Invalid",
                      Details: {
                          Name: "consolesamples.console.openshift.io",
                          Group: "apiextensions.k8s.io",
                          Kind: "CustomResourceDefinition",
                          UID: "",
                          Causes: [
                              {
                                  Type: "FieldValueForbidden",
                                  Message: "Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic",
                                  Field: "spec.validation.openAPIV3Schema.properties[spec].properties[tags].uniqueItems",
                              },
                          ],
                          RetryAfterSeconds: 0,
                      },
                      Code: 422,
                  },
              },
          },
      }
      unable to create CRD instances: unable to create CRD "consolesamples.console.openshift.io": CustomResourceDefinition.apiextensions.k8s.io "consolesamples.console.openshift.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[tags].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic
  occurred[0m
  [38;5;9mIn [1m[BeforeEach][0m[38;5;9m at: [1m/go/src/github.com/openshift/api/tests/generator.go:136[0m [38;5;243m@ 07/10/23 14:55:09.023[0m
```

From https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_api/1503/pull-ci-openshift-api-master-integration/1678413731614167040

I replaced the UniqueItems with `// +listType=set`, we used this a lot in the API.

Reply:
Can you add an integration test to prove that this works as expected?

Reply:
Added two tests:

  - name: Should decline a ConsoleSample when reusing a tag
  - name: Should decline a ConsoleSample with more then 10 tags